relay-server/src/processing/logs/mod.rs

```diff
@@ -0,0 +1,274 @@
+        }
+    }
+}
+
+/// Output produced by [`LogsProcessor`].
+#[expect(
+    clippy::large_enum_variant,
+    reason = "until we have a better solution to combat the excessive growth of Item, see #4819"
+)]
+pub enum LogOutput {
+    NotProcessed(Managed<SerializedLogs>),
+    Processed(Managed<ExpandedLogs>),
+}
+
+impl Forward for LogOutput {
+    fn serialize_envelope(self) -> Result<Managed<Box<Envelope>>, Rejected<()>> {
+        let logs = match self {
+            Self::NotProcessed(logs) => logs,
+            Self::Processed(logs) => logs.try_map(|logs, _| {
+                logs.serialize()
+                    .map_err(drop)
+                    .with_outcome(Outcome::Invalid(DiscardReason::Internal))
+            })?,
+        };
+
+        Ok(logs.map(|logs, _| logs.serialize_envelope()))
+    }
+}
+
+/// Logs in their serialized state, as transported in an envelope.
+pub struct SerializedLogs {
+    /// Original envelope headers.
+    headers: EnvelopeHeaders,
+
+    /// OTel Logs are not sent in containers, an envelope is very likely to contain multiple OTel logs.
+    otel_logs: Items,
+    /// Logs are sent in item containers, there is specified limit of a single container per
+    /// envelope.
+    ///
+    /// But at this point this has not yet been validated.
+    logs: Items,
+}
+
+impl SerializedLogs {
+    fn serialize_envelope(self) -> Box<Envelope> {
+        let mut items = self.logs;
+        items.extend(self.otel_logs);
+        Envelope::from_parts(self.headers, items)
+    }
+
+    fn items(&self) -> impl Iterator<Item = &Item> {
+        self.otel_logs.iter().chain(self.logs.iter())
+    }
+
+    /// Returns the total count of all logs contained.
+    ///
+    /// This contains all logical log items, not just envelope items and is safe
+    /// to use for rate limiting.
+    fn count(&self) -> usize {
+        self.items()
+            .map(|item| item.item_count().unwrap_or(1) as usize)
+            .sum()
+    }
+
+    /// Returns the sum of bytes of all logs contained.
+    fn bytes(&self) -> usize {
+        self.items().map(|item| item.len()).sum()
+    }
+}
+
+impl Counted for SerializedLogs {
+    fn quantities(&self) -> Quantities {
+        smallvec::smallvec![
+            (DataCategory::LogItem, self.count()),
+            (DataCategory::LogByte, self.bytes())
+        ]
+    }
+}
+
+impl RateLimited for Managed<SerializedLogs> {
+    type Error = Error;
+
+    async fn enforce<T>(
+        &mut self,
+        mut rate_limiter: T,
+        _ctx: Context<'_>,
+    ) -> Result<(), Rejected<Self::Error>>
+    where
+        T: RateLimiter,
+    {
+        let scoping = self.scoping();
+
+        let items = rate_limiter
+            .try_consume(scoping.item(DataCategory::LogItem), self.count())
+            .await;
+        let bytes = rate_limiter
+            .try_consume(scoping.item(DataCategory::LogByte), self.bytes())
+            .await;
+
+        let total_limits = items.merge_with(bytes);
```

Comment:
These are the limits that were breached, right? Perhaps
```suggestion
        let breached_limits = items.merge_with(bytes);
```